onnx/test/shape_inference_test.py

```diff
@@ -2850,6 +2850,104 @@ def test_einsum_incorrect_num_inputs(self):  # type: () -> None
             [])
         self.assertRaises(checker.ValidationError, self._inferred, graph)
 
+    def test_negative_log_likehood_shape_is_NCdd(self):  # type: () -> None
+        N, C = 3, 4
+        graph = self._make_graph(
+            [('input', TensorProto.FLOAT, (N, C)),
+             ('target', TensorProto.INT64, (N,))],
+            [make_node('NegativeLogLikelihoodLoss', ['input', 'target'], ['loss'], reduction='none')],
```

Comment:
Optional inputs can also be omitted by providing an empty string. I'm not sure if that case would currently be correctly handled or not?
```python
> node = onnx.helper.make_node(op_type="Nll", inputs=["a", "b", ""], outputs=["z"])
> len(node.input)
3
```

Reply:
replaced numInput/numOutput with hasInput(i)/hasOutput(i).